### PR TITLE
support for apfs date time

### PR DESCRIPTION
**Pull request by aguilajesus**
support for apfs date time

### CODE REVIEW

File: dfdatetime/apfs_time.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    number_of_days, hours, minutes, seconds = self._GetTimeValues(timestamp)
+
+    # trimming precision for consistency.
+    # TODO: fix string representation to reflect 9 digits precision.
+    microseconds = nanoseconds / definitions.MILLISECONDS_PER_SECOND
+
+    year, month, day_of_month = self._GetDateValuesWithEpoch(
+        number_of_days, self._EPOCH)
+
+    return '{0:04d}-{1:02d}-{2:02d} {3:02d}:{4:02d}:{5:02d}.{6:06d}'.format(
```
Comment (joachimmetz) [inline dfdatetime/apfs_time.py]: `{6:06d}` change to `{6:09d}`

File: dfdatetime/apfs_time.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    if (self._timestamp is None or self._timestamp < 0 or
+        self._timestamp > self._UINT64_MAX):
+      return None
+
+    timestamp, nanoseconds = divmod(
+        self._timestamp, definitions.NANOSECONDS_PER_SECOND)
+    number_of_days, hours, minutes, seconds = self._GetTimeValues(timestamp)
+
+    # trimming precision for consistency.
```
Comment (joachimmetz) [inline dfdatetime/apfs_time.py]: Please address this TODO

File: dfdatetime/apfs_time.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    self._timestamp = timestamp
+
+  def CopyToDateTimeString(self):
+    """Copies the APFS timestamp to a date and time string.
+
+    Returns:
+      str: date and time value formatted as: "YYYY-MM-DD hh:mm:ss.######" or
+          None if the timestamp is missing or invalid.
+    """
+    if (self._timestamp is None or self._timestamp < 0 or
```
Comment (joachimmetz) [inline dfdatetime/apfs_time.py]: Is the APFS timestamp unsigned ? I would expect time_t and there for the timestamp to be signed.

Please change to check for INT64_MIN and INT64_MAX bounds instead

File: tests/apfs_time.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    date_time_string = apfs_time_object.CopyToDateTimeString()
+    self.assertIsNone(date_time_string)
+
+  def testCopyToDateTimeStringISO8601(self):
+    """Tests the CopyToDateTimeStringISO8601 function."""
+    apfs_time_object = apfs_time.APFSTime(timestamp=1281643591987654321)
+
+    date_time_string = apfs_time_object.CopyToDateTimeStringISO8601()
+    self.assertEqual(date_time_string, '2010-08-12T20:06:31.987654Z')
```
Comment (joachimmetz) [inline tests/apfs_time.py]: Please use 9 digit precision for fraction of seconds

File: dfdatetime/apfs_time.py (hunk 3 above)
Comment (aguilajesus) [reply dfdatetime/apfs_time.py]: paper pointed in #116 claims is UINT64, I haven't been able to confirm this claim yet.

**Comment by joachimmetz** [comment]
Ok let's determine if the timestamp is actually unsigned. Seeing this would mean it would need to truncate date and time values before 1970.

**Comment by aguilajesus** [comment]
The upper band seems to match with UINT64 as well

```
touch -t 256001010000 time-test 

$ stat -x time-test
  File: "time-test"
  Size: 0            FileType: Regular File
  Mode: (0644/-rw-r--r--)         Uid: (1000/jesus)  Gid: (1000/jesus)
Device: 1,4   Inode: 7028872    Links: 1
Access: Fri Apr 11 16:47:16 2262
Modify: Fri Apr 11 16:47:16 2262
Change: Sun Sep 16 12:06:29 2018

stat -r time-test
<x> 7028872 0100644 1 1000 1000 0 0 **9223372036** **9223372036** 1537124789 0 4194304 0 0 time-test

python -c "print(divmod((1 << 63) - 1, 10**9))"
(**9223372036L**, 854775807L)
```

**Comment by aguilajesus** [comment]
Thanks for the review and nits on the missed checks

**Pull request merged**